internal/store/postgres_migrator.go

```diff
@@ -0,0 +1,80 @@
+package store
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	"gorm.io/gorm"
+)
+
+func tryLockExclusive(tx *gorm.DB) (bool, error) {
+	var success []bool
+	if err := tx.Raw("SELECT pg_try_advisory_lock(12345) as success").Pluck("success", &success).Error; err != nil {
+		return false, fmt.Errorf("failed to try exclusive advisory lock: %w", err)
+	}
+	if len(success) != 1 {
+		return false, fmt.Errorf("expected single result for exclusive lock. got %d", len(success))
+	}
+	return success[0], nil
+}
+
+func unlockExclusive(tx *gorm.DB) error {
+	var success []bool
+	if err := tx.Raw("SELECT pg_advisory_unlock(12345) as success").Pluck("success", &success).Error; err != nil {
+		return fmt.Errorf("failed to perform advisory unlock: %w", err)
+	}
+	if len(success) != 1 {
+		return fmt.Errorf("expected single result for exclusive unlock. got %d", len(success))
+	}
+	if !success[0] {
+		return errors.New("failed unlocking exclusive lock")
+	}
+	return nil
+}
+
+func lockShared(tx *gorm.DB) error {
+	if err := tx.Exec("SELECT pg_advisory_xact_lock_shared(12345)").Error; err != nil {
+		return fmt.Errorf("failed to perform advisory lock: %w", err)
+	}
+	return nil
+}
+
+type postgresMigrator struct {
+	db      *gorm.DB
+	migrate func() error
+	log     logrus.FieldLogger
+}
+
+func newPostgresMigrator(db *gorm.DB, migrate func() error, log logrus.FieldLogger) *postgresMigrator {
+	return &postgresMigrator{
+		db:      db,
+		migrate: migrate,
+		log:     log,
+	}
+}
+
+func (p *postgresMigrator) performMigrate() error {
+	// This transaction is leaked on purpose in order to keep the shared lock
+	// as long as the process is alive
+	tx := p.db.Begin()
```

Comment:
we leave the transaction open on purpose to avoid gorm to close it as part of its connection pooling

Reply:
Is there any details about this?

Why closing a transaction would close a connection?

Reply:
gorm has connection pooling https://gorm.io/docs/connecting_to_the_database.html#Connection-Pool .
The connection pool may close a connection if it is idle more than the connection lifetime set in the connection pool.  This shouldn't be a problem since every time a query is executed, a connection is taken from the pool or a new one is created.
There is an exception to this.  When using transaction, the same connection must be kept since they are all need to be part of the series of queries that are performed as a whole - meaning that they all need to be performed on the same database connection.
Since the user may want to continue queries on an open transaction, the connection pool cannot close it.

Reply:
But we would be closing the transaction after the .migrate

I still don't understand what you mean

tx: 
   lock()
   migrate
(tx is closed)

Reply:
The process is the following
```
    tx()
    lock_succeeded = try_lock_exclusive() // Only one should succeed
    lock_shared() // everyone should lock the shared lock
    if not lock_succeeded {
         return
    }
    // If we reached here then the process holds both exclusive lock and shared lock
    migrate()
    unlock_exclusive() // We are left with the shared lock
```

Reply:
So at the end everyone is left with a shared lock

Reply:
We decided to remove this part from the PR and possibly do it (or something similar) in another PR